public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx

```diff
@@ -409,6 +417,92 @@ export class VizPanelManager extends SceneObjectBase<VizPanelManagerState> {
     return this.state.panel.clone({ $data: this.state.$data?.clone() });
   }
 
+  public changeToLibraryPanel(libPanel: LibraryPanel) {
+    const panel = this.state.panel;
+    const vizPanel = this.state.sourcePanel.resolve();
+
+    // create the replacement libraryVizPanel
+    const libVizPanel = new LibraryVizPanel({
+      title: panel.state.title,
+      uid: libPanel.uid,
+      name: libPanel.name,
+      panelKey: panel.state.key!,
+    });
+
+    libVizPanel.activate();
+
+    // we need to update panel edit and dashboard with the new library panel after it's been loaded
+    libVizPanel.subscribeToState(() => this.updatePanelFromLibPanel(libPanel, libVizPanel, vizPanel));
```

Comment:
Do we need to update the dashboard? I mean - shouldn't this happen when the changes are commited? In the old architecture, after changing panel to a lib panel, one can still discard the changes, so I would assume we wait with the grid item replacement until commit.

Reply:
What do you mean by commit? As in saving or discarding the library panel in panel edit? My thinking was that selecting a lib panel from the pane is committing and after that, if there are any other changes to said library panel you have the options to save them or discard them in the context of updating the library panel and not the context of the dashboard.

In the old arch we can discard any panel, not only library panels. But in the new arch we do not have this option for all panels, we can only go back to dashboard, thus I assumed that the discard button (and also the save) for library panels refers to changes regarding saved lib panel template.

Do you think we should update the dashboard only when the library panel is saved, after selecting one from the lib panels pane?

Reply:
We should only update the dashboard from inside commit changes:

https://github.com/grafana/grafana/blob/main/public/app/features/dashboard-scene/panel-edit/PanelEditor.tsx#L104